flink-streaming-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/streaming/connectors/cassandra/CassandraSink.java

```
@@ -0,0 +1,313 @@
+	 * @param <IN>  input type
+	 * @return CassandraSinkBuilder, to further configure the sink
+	 */
+	public static <IN, T extends Tuple> CassandraSinkBuilder<IN> addSink(DataStream<IN> input) {
+		if (input.getType() instanceof TupleTypeInfo) {
+			DataStream<T> tupleInput = (DataStream<T>) input;
+			return (CassandraSinkBuilder<IN>) new CassandraTupleSinkBuilder<>(tupleInput, tupleInput.getType(), tupleInput.getType().createSerializer(tupleInput.getExecutionEnvironment().getConfig()));
+		} else {
+			return new CassandraPojoSinkBuilder<>(input, input.getType(), input.getType().createSerializer(input.getExecutionEnvironment().getConfig()));
+		}
+	}
+
+	public abstract static class CassandraSinkBuilder<IN> {
+		protected final DataStream<IN> input;
+		protected final TypeSerializer<IN> serializer;
+		protected final TypeInformation<IN> typeInfo;
+		protected ClusterBuilder builder;
+		protected String query;
+		protected CheckpointCommitter committer;
+		protected boolean isWriteAheadLogEnabled;
+
+		public CassandraSinkBuilder(DataStream<IN> input, TypeInformation<IN> typeInfo, TypeSerializer<IN> serializer) {
+			this.input = input;
+			this.typeInfo = typeInfo;
+			this.serializer = serializer;
+		}
+
+		/**
+		 * Sets the query that is to be executed for every record.
+		 *
+		 * @param query query to use
+		 * @return this builder
+		 */
+		public CassandraSinkBuilder<IN> setQuery(String query) {
+			this.query = query;
+			return this;
+		}
+
+		/**
+		 * Sets the cassandra host to connect to.
+		 *
+		 * @param host host to connect to
+		 * @return this builder
+		 */
+		public CassandraSinkBuilder<IN> setHost(String host) {
+			return setHost(host, 9042);
+		}
+
+		/**
+		 * Sets the cassandra host/port to connect to.
+		 *
+		 * @param host host to connect to
+		 * @param port port to connect to
+		 * @return this builder
+		 */
+		public CassandraSinkBuilder<IN> setHost(final String host, final int port) {
+			builder = new ClusterBuilder() {
```

I wonder whether we should throw an exception if the builder is already set. Otherwise, if somebody first sets a builder and then a host the results will be very misleading.